src/api/libcellml/importer.h

```diff
@@ -0,0 +1,169 @@
+ *
+ * The Importer class is for representing a CellML Importer.
+ */
+class LIBCELLML_EXPORT Importer: public Logger
+{
+public:
+    ~Importer() override; /**< Destructor */
+    Importer(const Importer &rhs) = delete; /**< Copy constructor */
+    Importer(Importer &&rhs) noexcept = delete; /**< Move constructor */
+    Importer &operator=(Importer rhs) = delete; /**< Assignment operator */
+
+    /**
+     * @brief Create a @c Importer object.
+     *
+     * Factory method to create a @c Importer.  Create a
+     * blank model with::
+     *
+     *   ImporterPtr importer = libcellml::Importer::create();
+     *
+     * @return A smart pointer to a @c Importer object.
+     */
+    static ImporterPtr create() noexcept;
+
+    /**
+     * @brief Flatten the @p model.
+     *
+     * Instantiates all imports and removes them from the @p model.
+     * The result is a self-contained model requiring no external
+     * resources and having no imports.
+     *
+     * @sa clone
+     */
+    ModelPtr flatten(const ModelPtr &inModel);
+
+    /**
+     * @brief Resolve all imports in the @p model.
+     *
+     * Resolve all @c Component and @c Units imports by loading the models
+     * from local disk through relative URLs.  The @p baseFile is used to determine
+     * the full path to the source model relative to this one.
+     * @param model The @c Model whose imports need resolution.
+     * @param baseFile The @c std::string location on local disk of the source @c Model.
+     */
+    void resolveImports(ModelPtr &model, const std::string &baseFile);
+
+    /**
+     * @brief Return the number of models present in the importer's library.
+     *
+     * @return The number of models present in the importer's library.
+     */
+    size_t libraryCount();
+
+    /**
+     * @brief Retrieve the @c ModelPtr instance from the importer library which was loaded
+     *        from the given @p key.
+     *
+     * @return A @c ModelPtr instance.
+     */
+    ModelPtr library(const std::string &key);
+
+    /**
+     * @brief Retrieve the @c ModelPtr instance from the importer library which was loaded
+     *        at the given index.
+     *
+     *  The @p index must be less than libraryCount().
+     *
+     * @return A @c ModelPtr instance.
+     */
+    ModelPtr library(const size_t &index);
+
+    /**
+     * @brief Manually add a local @c ModelPtr model instance to the importer library,
+     *        using the given @p key as a reference.
+     *
+     * If the given key already exists in the library, the function will return false
+     * and the library will not be changed.
+     * @sa replaceModel.
+     *
+     * @param model a @c ModelPtr instance to add.
+     * @param key a @c std::string representing the key to associate with the model.
+     *
+     * @return boolean value, @c true if the model was added, @c false if it was not.
+     */
+    bool addModel(const ModelPtr &model, const std::string &key);
```

Comment:
I don't remember the discussion, but... why do we need a key? As a user of the library, what key should I use? All I want is to add a model to my importer object.

Reply:
You need to give the model a handle by which it will be known - it needs a way to be identified to other models so that we can import from this one.  When using files, it's the filename and path, or absolute URL.  Inside the importer, it can be a string (or a URL too, if you want).

Reply:
Ok, makes sense now. I am just not sure about the use of `key`. As you wrote it, it's to identify the model to other models. So, I would expect the parameter to be called `identifier` or `id` for short. Now, I know that `id` is used in the context of annotations, but `key` really seems wrong to me.

Reply:
It's because the library is basically a map, and maps have a unique string key and a value.  In this case, the key is either user-defined or the URL, and the value is the model.  I don't like having id/identifier, as that's used elsewhere as you say.

Reply:
Hmm... then, shouldn't `addModel()` and `replaceModel()` be replaced with `addLibrary()` and `replaceLibrary()`? Also, for logical reasons, I would have `addLibrary(key, model)` and `replaceLibrary(key, model)`, i.e. swap `key` and `model`.